components/org.wso2.carbon.identity.user.endpoint/src/main/java/org/wso2/carbon/identity/user/endpoint/impl/ValidateUsernameApiServiceImpl.java

```diff
@@ -78,27 +83,31 @@ public Response validateUsernamePost(UsernameValidationRequestDTO user) {
-                        tenantDomain, SelfRegistrationStatusCodes.ERROR_CODE_SELF_REGISTRATION_DISABLED));
+                logDebug(String.format("Self registration is not enabled for tenant domain : %s . Hence returning " +
+                                "code %s", tenantDomain,
+                        SelfRegistrationStatusCodes.ERROR_CODE_SELF_REGISTRATION_DISABLED));
                 responseDTO.setStatusCode(
                         Integer.parseInt(SelfRegistrationStatusCodes.ERROR_CODE_SELF_REGISTRATION_DISABLED));
-            } else if (userSelfRegistrationManager.isUsernameAlreadyTaken(user.getUsername(), tenantDomain)) {
+            } else if (userSelfRegistrationManager.isUsernameAlreadyTaken(fullyQualifiedUsername, tenantDomain)) {
                 logDebug(String.format("username : %s is an already taken. Hence returning code %s: ",
-                        user.getUsername(), SelfRegistrationStatusCodes.ERROR_CODE_USER_ALREADY_EXISTS));
+                        fullyQualifiedUsername, SelfRegistrationStatusCodes.ERROR_CODE_USER_ALREADY_EXISTS));
                 responseDTO.setStatusCode(Integer.parseInt(SelfRegistrationStatusCodes.ERROR_CODE_USER_ALREADY_EXISTS));
-            } else if (!userSelfRegistrationManager.isMatchUserNameRegex(tenantDomain, user.getUsername())) {
+            } else if (!userSelfRegistrationManager.isMatchUserNameRegex(tenantDomain, username)) {
                 logDebug(String.format("%s is an invalid user name. Hence returning code %s: ",
-                        user.getUsername(), SelfRegistrationStatusCodes.CODE_USER_NAME_INVALID));
+                        fullyQualifiedUsername, SelfRegistrationStatusCodes.CODE_USER_NAME_INVALID));
                 responseDTO.setStatusCode(Integer.parseInt(SelfRegistrationStatusCodes.CODE_USER_NAME_INVALID));
             } else {
-                logDebug(String.format("username : %s is avilable for self registration. Hence returning code %s: ",
-                        user.getUsername(), SelfRegistrationStatusCodes.CODE_USER_NAME_AVAILABLE));
+                logDebug(String.format("username : %s is available for self registration. Hence returning code %s: ",
+                        fullyQualifiedUsername, SelfRegistrationStatusCodes.CODE_USER_NAME_AVAILABLE));
                 responseDTO.setStatusCode(Integer.parseInt(SelfRegistrationStatusCodes.CODE_USER_NAME_AVAILABLE));
             }
             return Response.ok().entity(responseDTO).build();
         } catch (IdentityRecoveryException e) {
             if (LOG.isDebugEnabled()) {
-                LOG.debug("Error while checking username validity for user " + user.getUsername(), e);
+                LOG.debug("Error while checking username validity for user " + fullyQualifiedUsername, e);
+            }
+            if (e instanceof IdentityRecoveryClientException && StringUtils.isNotBlank(e.getErrorDescription())) {
```

Comment:
Can we send a more general error client error when the description of the exception is blank? I think we can handle it via a separate catch block. WDYT?

Reply:
IMO if we send a `Response.Status.BAD_REQUEST` without proper ErrorDescription it won't be meaning full to the end user. ATST aren't we expecting an error description for all IdentityRecoveryClientExceptions? if the exception is not well defined then it is better show `Response.Status.INTERNAL_SERVER_ERROR`. WDTY?

Reply:
We are expecting an error message in `IdentityRecoveryClientExceptions`. But we need to handle the scenarios where the message can be blank. The purpose of the `IdentityRecoveryClientExceptions` is to represent an error in the request. `INTERNAL_SERVER_ERROR` is for errors in the server. So if we send the `INTERNAL_SERVER_ERROR` for a `BAD_REQUEST ` scenario, it conveys that server has a problem, not the request. Therefore, user will contiue to send the same request over and over again without correcting the request.

So we can send a general message such as `Invalid request` etc. WDYT?